src/state.rs

```diff
@@ -0,0 +1,295 @@
+        self.get_current_package().resolve(symbol_name)
+    }
+
+    pub fn intern<A: AsRef<str>>(&mut self, symbol_name: A) -> SymbolRef {
+        self.get_current_package_mut()
+            .intern(String::from(symbol_name.as_ref()))
+    }
+
+    pub fn import(&mut self, symbols: &[SymbolRef]) -> Result<()> {
+        self.get_current_package_mut().import(symbols)
+    }
+
+    pub fn use_package(&mut self, package: &Package) -> Result<()> {
+        self.get_current_package_mut().use_package(package)
+    }
+
+    pub fn fmt_to_string(&self, symbol: &SymbolRef) -> String {
+        self.get_current_package().fmt_to_string(symbol)
+    }
+
+    fn intern_fold<A: AsRef<str>>(
+        &mut self,
+        init: SymbolRef,
+        path: &[A],
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        path.iter()
+            .try_fold(init, |acc, s| match self.symbol_packages.get_mut(&acc) {
+                Some(package) => Ok(package.intern(String::from(s.as_ref()))),
+                None => {
+                    if create_unknown_packges {
+                        let mut package = Package::new(acc);
+                        let symbol = package.intern(String::from(s.as_ref()));
+                        self.add_package(package);
+                        Ok(symbol)
+                    } else {
+                        bail!("Package {acc} not found")
+                    }
+                }
+            })
+    }
+
+    #[inline]
+    pub fn intern_path<A: AsRef<str>>(
+        &mut self,
+        path: &[A],
+        keyword: bool,
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        self.intern_fold(Symbol::root(keyword).into(), path, create_unknown_packges)
+    }
+
+    #[inline]
+    pub fn intern_relative_path<A: AsRef<str>>(
+        &mut self,
+        path: &[A],
+        create_unknown_packges: bool,
+    ) -> Result<SymbolRef> {
+        self.intern_fold(self.current_package.clone(), path, create_unknown_packges)
+    }
+
+    #[inline]
+    pub fn minimal() -> Self {
+        Self::new_with_package(Package::new(SymbolRef::new(Symbol::root_sym())))
+    }
+
+    pub fn init_lurk_state() -> Self {
+        let mut root_package = Package::new(SymbolRef::new(Symbol::root_sym()));
+
+        // bootstrap the keyword package
+        let keyword_package = Package::new(SymbolRef::new(Symbol::root_key()));
+
+        // bootstrap the lurk package
+        let mut lurk_package = Package::new(root_package.intern(LURK_PACKAGE_SYMBOL_NAME.into()));
+        LURK_PACKAGE_SYMBOLS_NAMES.iter().for_each(|symbol_name| {
+            lurk_package.intern(symbol_name.to_string());
+        });
+
+        // bootstrap the meta package
+        let mut meta_package = Package::new(lurk_package.intern(META_PACKAGE_SYMBOL_NAME.into()));
+        META_PACKAGE_SYMBOLS_NAMES.iter().for_each(|symbol_name| {
+            meta_package.intern(symbol_name.to_string());
+        });
```

Comment:
I think there is a `defpackage` function trying to be born here. Eventually `defpackage` will likely be a macro, so if you want to name something like `define_package` to signal the functional helper that will underly it, that would be fine. As a nice side effect, the REPL's `!(defpackage …)` can be more powerful more or less for free.

Reply:
Yes, in the future this function might be a specialized call to the function that does `defpackage`